File: src/lib/medical-safety.ts
export type SafetySeverity = 'error' | 'warning';

export interface SafetyIssue {
    code: string;
    severity: SafetySeverity;
    message: string;
    evidence?: string;
}

export interface SafetyReport {
    verdict: 'PASS' | 'NEEDS_REVIEW' | 'FAIL';
    issues: SafetyIssue[];
    checks: {
        placeholderFree: boolean;
        guidelineCitationsValid: boolean;
        pubMedCitationsValid: boolean;
        quotedClaimsGrounded: boolean;
        numericClaimsGrounded: boolean;
    };
}

export interface SafetySource {
    text: string;
    label: string;
}

export interface RunSafetyChecksParams {
    letter: string;
    clinicalNotes: string;
    denialReason?: string;
    guidelineSources?: SafetySource[];
    pubMedSources?: SafetySource[];
}

const ALLOWED_BRACKET_CONTENT = /^(?:\d+|PubMed\s+\d+|Reference\s+\d+|DOCUMENTATION GAP(?::|\s|$).*)$/i;
const PLACEHOLDER_HINTS = /\b(?:address|phone|fax|date|dob|birth|member|claim|insert|if known|not provided|omitted|current|provider|physician|clinician|doctor|letterhead|practice|facility|office|npi|name|signature|sincerely|to be|blank)\b/i;
const CITATION_OR_SECTION_NUMBER = /^\d{1,2}$/;
const HEADING_ONLY_QUOTE = /^(?:active\s+)?(?:diagnosis|diagnoses|medications?|problems?|assessment|plan|impression|history|clinical summary|review of systems|physical examination|objective data|laboratory results|supporting documentation|documentation gaps?)$/i;

function normalizeText(value: string): string {
    return value
        .normalize('NFKC')
        .replace(/[“”]/g, '"')
        .replace(/[‘’]/g, "'")
        .replace(/[–—−]/g, '-')
        .replace(/\s+/g, ' ')
        .trim()
        .toLowerCase();
}

function normalizeForGrounding(value: string): string {
    return normalizeText(value)
        .replace(/&/g, ' and ')
        .replace(/[^a-z0-9%/]+/g, ' ')
        .replace(/\s+/g, ' ')
        .trim();
}

function groundingTokens(value: string): string[] {
    return normalizeForGrounding(value)
        .split(' ')
        .filter(token => token.length > 1 || /\d/.test(token));
}

function hasOrderedTokenWindow(quote: string, source: string): boolean {
    const quoteTokens = groundingTokens(quote);
    const sourceTokens = groundingTokens(source);
    if (quoteTokens.length < 4 || sourceTokens.length < quoteTokens.length) return false;

    const maxWindow = quoteTokens.length + Math.max(6, Math.ceil(quoteTokens.length * 0.75));
    for (let start = 0; start < sourceTokens.length; start += 1) {
        if (sourceTokens[start] !== quoteTokens[0]) continue;
        let quoteIndex = 0;
        const end = Math.min(sourceTokens.length, start + maxWindow);
        for (let sourceIndex = start; sourceIndex < end; sourceIndex += 1) {
            if (sourceTokens[sourceIndex] === quoteTokens[quoteIndex]) {
                quoteIndex += 1;
                if (quoteIndex === quoteTokens.length) return true;
            }
        }
    }

    return false;
}

function unique<T>(values: T[]): T[] {
    return [...new Set(values)];
}

function sourceContains(value: string, sources: SafetySource[]): boolean {
    const normalized = normalizeForGrounding(value);
    if (!normalized) return true;
    return sources.some(source => {
        const normalizedSource = normalizeForGrounding(source.text);
        return normalizedSource.includes(normalized) || hasOrderedTokenWindow(value, source.text);
    });
}

function findBracketIssues(letter: string): SafetyIssue[] {
    const issues: SafetyIssue[] = [];
    const matches = letter.matchAll(/\[([^\]]+)\]/g);

    for (const match of matches) {
        const content = match[1].trim();
        if (ALLOWED_BRACKET_CONTENT.test(content)) continue;

        if (PLACEHOLDER_HINTS.test(content) || !/^\d+$/.test(content)) {
            issues.push({
                code: 'UNRESOLVED_PLACEHOLDER',
                severity: 'error',
                message: `Potential unresolved placeholder or unsupported bracketed text: [${content}]`,
                evidence: `[${content}]`,
            });
        }
    }

    return issues;
}

function isPlaceholderBracket(content: string): boolean {
    if (ALLOWED_BRACKET_CONTENT.test(content)) return false;
    return PLACEHOLDER_HINTS.test(content) || /^[A-Z0-9 _/-]{3,}$/.test(content);
}

export function sanitizeGeneratedLetter(letter: string): string {
    return letter
        .replace(/\s*\[SOURCE [A-Z]\]/g, '')
        .replace(/^(\s{0,3}(?:#{1,6}\s*)?(?:\*\*)?)Documentation Gaps(?:\*\*)?\s*$/gim, '$1Clinical Rationale for Exception')
        .replace(/\[DOCUMENTATION GAP[.:]\s*([\s\S]*?)\]/g, 'Additional supporting rationale: $1')
        .replace(/\[DOCUMENTATION GAP\]\s*:\s*/g, 'Additional supporting rationale: ')
        .replace(/\[DOCUMENTATION GAP[.:]\s*/g, 'Additional supporting rationale: ')
        .split('\n')
        .map(line => {
            const bracketMatches = [...line.matchAll(/\[([^\]]+)\]/g)];
            if (
                bracketMatches.length > 0
                && bracketMatches.every(match => isPlaceholderBracket(match[1].trim()))
            ) {
                return '';
            }
            return line.replace(/\[([^\]]+)\]/g, (full, content: string) => (
                isPlaceholderBracket(content.trim()) ? '' : full
            ));
        })
        .join('\n')
        .replace(/[ \t]+\n/g, '\n')
        .replace(/\n{3,}/g, '\n\n')
        .trim();
}

function findGuidelineCitationIssues(letter: string, guidelineCount: number): SafetyIssue[] {
    const issues: SafetyIssue[] = [];
    const numericRefs = unique(
        [...letter.matchAll(/(?<!PubMed\s)\[(\d+)\]/gi)].map(match => Number(match[1]))
    );

    for (const ref of numericRefs) {
        if (ref < 1 || ref > guidelineCount) {
            issues.push({
                code: 'INVALID_GUIDELINE_CITATION',
                severity: 'error',
                message: `Guideline citation [${ref}] does not map to any retrieved guideline source.`,
                evidence: `[${ref}]`,
            });
        }
    }

    if (guidelineCount > 0 && numericRefs.length === 0) {
        issues.push({
            code: 'MISSING_GUIDELINE_CITATIONS',
            severity: 'warning',
            message: 'Letter includes guideline context but does not cite any numbered guideline references.',
        });
    }

    return issues;
}

function findPubMedCitationIssues(letter: string, pubMedCount: number): SafetyIssue[] {
    const issues: SafetyIssue[] = [];
    const pubMedRefs = unique(
        [...letter.matchAll(/\[PubMed\s+(\d+)\]/gi)].map(match => Number(match[1]))
    );

    for (const ref of pubMedRefs) {
        if (ref < 1 || ref > pubMedCount) {
            issues.push({
                code: 'INVALID_PUBMED_CITATION',
                severity: 'error',
                message: `PubMed citation [PubMed ${ref}] does not map to any retrieved PubMed source.`,
                evidence: `[PubMed ${ref}]`,
            });
        }
    }

    return issues;
}

function quotedClaimsNeedGrounding(value: string): boolean {
    if (value.length < 8) return false;
    if (/^https?:\/\//i.test(value)) return false;
    if (HEADING_ONLY_QUOTE.test(value.trim())) return false;
    return /\d|pain|mass|lesion|deficit|stenosis|fracture|cancer|tumor|malign|biopsy|therapy|medication|exam|imaging|mri|ct|pet|x-ray|xray|lab|a1c|bmi|symptom|diagnos|criterion|requires|denial|not medically necessary/i.test(value);
}

function findQuotedClaimIssues(letter: string, sources: SafetySource[]): SafetyIssue[] {
    const issues: SafetyIssue[] = [];
    const quoted = unique(
        [...letter.matchAll(/"([^"]{4,240})"/g)]
            .map(match => match[1].trim())
            .filter(quotedClaimsNeedGrounding)
    );

    for (const quote of quoted) {
        if (!sourceContains(quote, sources)) {
            issues.push({
                code: 'UNGROUNDED_QUOTE',
                severity: 'warning',
                message: 'Quoted clinical, denial, or guideline text was not found verbatim in the supplied sources.',
                evidence: quote,
            });
        }
    }

    return issues;
}

function extractNumericFacts(text: string): string[] {
    const facts = [...text.matchAll(/\b(?:\d{1,2}[/-]\d{1,2}[/-]\d{2,4}|\d+(?:\.\d+)?\s?(?:%|mg\/dL|mmHg|cm|mm|kg\/m2|kg\/m²|weeks?|months?|years?|days?|hours?|\/10|points?|QALY|mL|L))\b/gi)]
        .map(match => match[0].trim())
        .filter(value => !CITATION_OR_SECTION_NUMBER.test(value));

    return unique(facts);
}

function findNumericClaimIssues(letter: string, sources: SafetySource[]): SafetyIssue[] {
    const issues: SafetyIssue[] = [];
    const numbers = extractNumericFacts(letter);

    for (const value of numbers) {
        if (!sourceContains(value, sources)) {
            issues.push({
                code: 'UNGROUNDED_NUMERIC_CLAIM',
                severity: 'warning',
                message: 'Numeric value in the letter was not found in the clinical notes, guidelines, denial reason, or PubMed context.',
                evidence: value,
            });
        }
    }

    return issues;
}

function buildSources(params: RunSafetyChecksParams): SafetySource[] {
    return [
        { label: 'clinicalNotes', text: params.clinicalNotes },
        { label: 'denialReason', text: params.denialReason || '' },
        ...(params.guidelineSources || []),
        ...(params.pubMedSources || []),
    ];
}

export function runMedicalSafetyChecks(params: RunSafetyChecksParams): SafetyReport {
    const guidelineCount = params.guidelineSources?.length || 0;
    const pubMedCount = params.pubMedSources?.length || 0;
    const allSources = buildSources(params);

    const placeholderIssues = findBracketIssues(params.letter);
    const guidelineCitationIssues = findGuidelineCitationIssues(params.letter, guidelineCount);
    const pubMedCitationIssues = findPubMedCitationIssues(params.letter, pubMedCount);
    const quotedClaimIssues = findQuotedClaimIssues(params.letter, allSources);
    const numericClaimIssues = findNumericClaimIssues(params.letter, allSources);

    const issues = [
        ...placeholderIssues,
        ...guidelineCitationIssues,
        ...pubMedCitationIssues,
        ...quotedClaimIssues,
        ...numericClaimIssues,
    ];

    const hasErrors = issues.some(issue => issue.severity === 'error');
    const hasWarnings = issues.some(issue => issue.severity === 'warning');

    return {
        verdict: hasErrors ? 'FAIL' : hasWarnings ? 'NEEDS_REVIEW' : 'PASS',
        issues,
        checks: {
            placeholderFree: placeholderIssues.length === 0,
            guidelineCitationsValid: guidelineCitationIssues.every(issue => issue.severity !== 'error'),
            pubMedCitationsValid: pubMedCitationIssues.length === 0,
            quotedClaimsGrounded: quotedClaimIssues.length === 0,
            numericClaimsGrounded: numericClaimIssues.length === 0,
        },
    };
}

export function safetyReportToVerificationChecks(report: SafetyReport) {
    return [
        {
            check: 'Deterministic Placeholder Check',
            status: report.checks.placeholderFree ? 'PASS' : 'FAIL',
            details: report.checks.placeholderFree
                ? 'No unresolved bracket placeholders were detected.'
                : report.issues.filter(i => i.code === 'UNRESOLVED_PLACEHOLDER').map(i => i.message).join(' '),
        },
        {
            check: 'Deterministic Citation Check',
            status: report.checks.guidelineCitationsValid && report.checks.pubMedCitationsValid ? 'PASS' : 'FAIL',
            details: report.checks.guidelineCitationsValid && report.checks.pubMedCitationsValid
                ? 'All numbered guideline and PubMed citations map to retrieved sources.'
                : report.issues.filter(i => i.code.includes('CITATION')).map(i => i.message).join(' '),
        },
        {
            check: 'Deterministic Grounding Check',
            status: report.checks.quotedClaimsGrounded && report.checks.numericClaimsGrounded ? 'PASS' : 'WARNING',
            details: report.checks.quotedClaimsGrounded && report.checks.numericClaimsGrounded
                ? 'Quoted and numeric claims were found in the supplied sources.'
                : report.issues.filter(i => i.code.includes('UNGROUNDED')).map(i => `${i.message} ${i.evidence || ''}`).join(' '),
        },
    ] as const;
}

File: backend/app/services/medical_safety.py
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass
from typing import Literal


SafetySeverity = Literal["error", "warning"]


@dataclass(frozen=True)
class SafetySource:
    label: str
    text: str


ALLOWED_BRACKET_CONTENT = re.compile(
    r"^(?:\d+|PubMed\s+\d+|Reference\s+\d+|DOCUMENTATION GAP(?::|\s|$).*)$",
    re.IGNORECASE,
)
PLACEHOLDER_HINTS = re.compile(
    r"\b(?:address|phone|fax|date|dob|birth|member|claim|insert|if known|not provided|"
    r"omitted|current|provider|physician|clinician|doctor|letterhead|practice|facility|"
    r"office|npi|name|signature|sincerely|to be|blank)\b",
    re.IGNORECASE,
)
CITATION_OR_SECTION_NUMBER = re.compile(r"^\d{1,2}$")
HEADING_ONLY_QUOTE = re.compile(
    r"^(?:active\s+)?(?:diagnosis|diagnoses|medications?|problems?|assessment|plan|"
    r"impression|history|clinical summary|review of systems|physical examination|"
    r"objective data|laboratory results|supporting documentation|documentation gaps?)$",
    re.IGNORECASE,
)


def _normalize_text(value: str) -> str:
    return (
        unicodedata.normalize("NFKC", value)
        .replace("“", '"')
        .replace("”", '"')
        .replace("‘", "'")
        .replace("’", "'")
        .replace("–", "-")
        .replace("—", "-")
        .replace("−", "-")
        .replace("²", "2")
        .replace("³", "3")
        .lower()
    )


def _normalize_for_grounding(value: str) -> str:
    text = _normalize_text(value).replace("&", " and ")
    text = re.sub(r"[^a-z0-9%/]+", " ", text)
    return re.sub(r"\s+", " ", text).strip()


def _grounding_tokens(value: str) -> list[str]:
    return [
        token
        for token in _normalize_for_grounding(value).split()
        if len(token) > 1 or re.search(r"\d", token)
    ]


def _has_ordered_token_window(quote: str, source: str) -> bool:
    quote_tokens = _grounding_tokens(quote)
    source_tokens = _grounding_tokens(source)
    if len(quote_tokens) < 4 or len(source_tokens) < len(quote_tokens):
        return False

    max_window = len(quote_tokens) + max(6, (3 * len(quote_tokens) + 3) // 4)
    for start, token in enumerate(source_tokens):
        if token != quote_tokens[0]:
            continue
        quote_index = 0
        for source_token in source_tokens[start : start + max_window]:
            if source_token == quote_tokens[quote_index]:
                quote_index += 1
                if quote_index == len(quote_tokens):
                    return True
    return False


def _source_contains(value: str, sources: list[SafetySource]) -> bool:
    normalized = _normalize_for_grounding(value)
    if not normalized:
        return True
    return any(
        normalized in _normalize_for_grounding(source.text)
        or _has_ordered_token_window(value, source.text)
        for source in sources
    )


def _unique(values: list[str]) -> list[str]:
    return list(dict.fromkeys(values))


def _find_bracket_issues(letter: str) -> list[dict]:
    issues: list[dict] = []
    for match in re.finditer(r"\[([^\]]+)\]", letter):
        content = match.group(1).strip()
        if ALLOWED_BRACKET_CONTENT.match(content):
            continue
        if PLACEHOLDER_HINTS.search(content) or not content.isdigit():
            issues.append(
                {
                    "code": "UNRESOLVED_PLACEHOLDER",
                    "severity": "error",
                    "message": f"Potential unresolved placeholder or unsupported bracketed text: [{content}]",
                    "evidence": f"[{content}]",
                }
            )
    return issues


def _is_placeholder_bracket(content: str) -> bool:
    if ALLOWED_BRACKET_CONTENT.match(content):
        return False
    return bool(PLACEHOLDER_HINTS.search(content) or re.match(r"^[A-Z0-9 _/-]{3,}$", content))


def sanitize_generated_letter(letter: str) -> str:
    text = re.sub(r"\s*\[SOURCE [A-Z]\]", "", letter)
    text = re.sub(
        r"^\s{0,3}(?:#{1,6}\s*)?(?:\*\*)?SECTION\s+\d+\s*[—\-:]\s*(?:HEADER\s*&\s*IDENTIFICATION|PURPOSE)(?:\*\*)?\s*$",
        "",
        text,
        flags=re.I | re.M,
    )
    text = re.sub(
        r"^\s{0,3}(?:#{1,6}\s*)?(?:\*\*)?SECTION\s+\d+\s*[—\-:]\s*",
        "",
        text,
        flags=re.I | re.M,
    )
    text = re.sub(
        r"^(\s{0,3}(?:#{1,6}\s*)?(?:\*\*)?)Documentation Gaps(?:\*\*)?\s*$",
        r"\1Clinical Rationale for Exception",
        text,
        flags=re.I | re.M,
    )
    text = re.sub(
        r"\[DOCUMENTATION GAP[.:]\s*(.*?)\]",
        r"Additional supporting rationale: \1",
        text,
        flags=re.S,
    )
    text = re.sub(r"\[DOCUMENTATION GAP\]\s*:\s*", "Additional supporting rationale: ", text)
    text = re.sub(r"\[DOCUMENTATION GAP[.:]\s*", "Additional supporting rationale: ", text)
    cleaned_lines: list[str] = []
    for line in text.splitlines():
        bracket_matches = list(re.finditer(r"\[([^\]]+)\]", line))
        if bracket_matches and all(_is_placeholder_bracket(m.group(1).strip()) for m in bracket_matches):
            cleaned_lines.append("")
            continue

        cleaned_lines.append(
            re.sub(
                r"\[([^\]]+)\]",
                lambda m: "" if _is_placeholder_bracket(m.group(1).strip()) else m.group(0),
                line,
            )
        )

    return re.sub(r"\n{3,}", "\n\n", "\n".join(cleaned_lines)).strip()


def _find_guideline_citation_issues(letter: str, guideline_count: int) -> list[dict]:
    issues: list[dict] = []
    numeric_refs = _unique(
        [
            match.group(1)
            for match in re.finditer(r"\[(\d+)\]", letter)
            if letter[max(0, match.start() - 10) : match.start()].lower().find("pubmed") == -1
        ]
    )

    for ref_text in numeric_refs:
        ref = int(ref_text)
        if ref < 1 or ref > guideline_count:
            issues.append(
                {
                    "code": "INVALID_GUIDELINE_CITATION",
                    "severity": "error",
                    "message": f"Guideline citation [{ref}] does not map to any retrieved guideline source.",
                    "evidence": f"[{ref}]",
                }
            )

    if guideline_count > 0 and not numeric_refs:
        issues.append(
            {
                "code": "MISSING_GUIDELINE_CITATIONS",
                "severity": "warning",
                "message": "Letter includes guideline context but does not cite any numbered guideline references.",
            }
        )

    return issues


def _find_pubmed_citation_issues(letter: str, pubmed_count: int) -> list[dict]:
    issues: list[dict] = []
    refs = _unique([match.group(1) for match in re.finditer(r"\[PubMed\s+(\d+)\]", letter, re.I)])
    for ref_text in refs:
        ref = int(ref_text)
        if ref < 1 or ref > pubmed_count:
            issues.append(
                {
                    "code": "INVALID_PUBMED_CITATION",
                    "severity": "error",
                    "message": f"PubMed citation [PubMed {ref}] does not map to any retrieved PubMed source.",
                    "evidence": f"[PubMed {ref}]",
                }
            )
    return issues


def _quoted_claim_needs_grounding(value: str) -> bool:
    if len(value) < 8:
        return False
    if re.match(r"^https?://", value, re.I):
        return False
    if HEADING_ONLY_QUOTE.match(value.strip()):
        return False
    return bool(
        re.search(
            r"\d|pain|mass|lesion|deficit|stenosis|fracture|cancer|tumor|malign|biopsy|"
            r"therapy|medication|exam|imaging|mri|ct|pet|x-ray|xray|lab|a1c|bmi|symptom|"
            r"diagnos|criterion|requires|denial|not medically necessary",
            value,
            re.I,
        )
    )


def _find_quoted_claim_issues(letter: str, sources: list[SafetySource]) -> list[dict]:
    issues: list[dict] = []
    quoted = _unique(
        [
            match.group(1).strip()
            for match in re.finditer(r'"([^"]{4,240})"', letter)
            if _quoted_claim_needs_grounding(match.group(1).strip())
        ]
    )
    for quote in quoted:
        if not _source_contains(quote, sources):
            issues.append(
                {
                    "code": "UNGROUNDED_QUOTE",
                    "severity": "warning",
                    "message": "Quoted clinical, denial, or guideline text was not found verbatim in the supplied sources.",
                    "evidence": quote,
                }
            )
    return issues


def _extract_numeric_facts(text: str) -> list[str]:
    facts = [
        match.group(0).strip()
        for match in re.finditer(
            r"\b(?:\d{1,2}[/-]\d{1,2}[/-]\d{2,4}|\d+(?:\.\d+)?\s?(?:%|mg/dL|mmHg|"
            r"cm|mm|kg/m2|kg/m²|weeks?|months?|years?|days?|hours?|/10|points?|"
            r"QALY|mL|L))\b",
            text,
            re.I,
        )
    ]
    return _unique([value for value in facts if not CITATION_OR_SECTION_NUMBER.match(value)])


def _find_numeric_claim_issues(letter: str, sources: list[SafetySource]) -> list[dict]:
    issues: list[dict] = []
    for value in _extract_numeric_facts(letter):
        if not _source_contains(value, sources):
            issues.append(
                {
                    "code": "UNGROUNDED_NUMERIC_CLAIM",
                    "severity": "warning",
                    "message": "Numeric value in the letter was not found in the clinical notes, guidelines, denial reason, or PubMed context.",
                    "evidence": value,
                }
            )
    return issues


def run_medical_safety_checks(
    *,
    letter: str,
    clinical_notes: str,
    denial_reason: str = "",
    guideline_sources: list[SafetySource] | None = None,
    pubmed_sources: list[SafetySource] | None = None,
) -> dict:
    guideline_sources = guideline_sources or []
    pubmed_sources = pubmed_sources or []
    all_sources = [
        SafetySource(label="clinicalNotes", text=clinical_notes),
        SafetySource(label="denialReason", text=denial_reason),
        *guideline_sources,
        *pubmed_sources,
    ]

    placeholder_issues = _find_bracket_issues(letter)
    guideline_citation_issues = _find_guideline_citation_issues(letter, len(guideline_sources))
    pubmed_citation_issues = _find_pubmed_citation_issues(letter, len(pubmed_sources))
    quoted_claim_issues = _find_quoted_claim_issues(letter, all_sources)
    numeric_claim_issues = _find_numeric_claim_issues(letter, all_sources)

    issues = [
        *placeholder_issues,
        *guideline_citation_issues,
        *pubmed_citation_issues,
        *quoted_claim_issues,
        *numeric_claim_issues,
    ]
    has_errors = any(issue["severity"] == "error" for issue in issues)
    has_warnings = any(issue["severity"] == "warning" for issue in issues)

    return {
        "verdict": "FAIL" if has_errors else "NEEDS_REVIEW" if has_warnings else "PASS",
        "issues": issues,
        "checks": {
            "placeholderFree": not placeholder_issues,
            "guidelineCitationsValid": not any(
                issue["severity"] == "error" for issue in guideline_citation_issues
            ),
            "pubMedCitationsValid": not pubmed_citation_issues,
            "quotedClaimsGrounded": not quoted_claim_issues,
            "numericClaimsGrounded": not numeric_claim_issues,
        },
    }


def safety_report_to_verification_checks(report: dict) -> list[dict]:
    checks = report["checks"]
    return [
        {
            "check": "Deterministic Placeholder Check",
            "status": "PASS" if checks["placeholderFree"] else "FAIL",
            "details": (
                "No unresolved bracket placeholders were detected."
                if checks["placeholderFree"]
                else " ".join(
                    issue["message"]
                    for issue in report["issues"]
                    if issue["code"] == "UNRESOLVED_PLACEHOLDER"
                )
            ),
        },
        {
            "check": "Deterministic Citation Check",
            "status": (
                "PASS"
                if checks["guidelineCitationsValid"] and checks["pubMedCitationsValid"]
                else "FAIL"
            ),
            "details": (
                "All numbered guideline and PubMed citations map to retrieved sources."
                if checks["guidelineCitationsValid"] and checks["pubMedCitationsValid"]
                else " ".join(
                    issue["message"] for issue in report["issues"] if "CITATION" in issue["code"]
                )
            ),
        },
        {
            "check": "Deterministic Grounding Check",
            "status": (
                "PASS"
                if checks["quotedClaimsGrounded"] and checks["numericClaimsGrounded"]
                else "WARNING"
            ),
            "details": (
                "Quoted and numeric claims were found in the supplied sources."
                if checks["quotedClaimsGrounded"] and checks["numericClaimsGrounded"]
                else " ".join(
                    f"{issue['message']} {issue.get('evidence', '')}"
                    for issue in report["issues"]
                    if "UNGROUNDED" in issue["code"]
                )
            ),
        },
    ]


def split_guideline_sources(rag_context: str) -> list[SafetySource]:
    if not rag_context.strip():
        return []
    matches = list(re.finditer(r"\[Reference\s+(\d+)\]", rag_context, re.I))
    if not matches:
        return [SafetySource(label="Guideline Context", text=rag_context)]
    sources: list[SafetySource] = []
    for index, match in enumerate(matches):
        end = matches[index + 1].start() if index + 1 < len(matches) else len(rag_context)
        sources.append(SafetySource(label=f"Reference {match.group(1)}", text=rag_context[match.start() : end]))
    return sources


def split_pubmed_sources(context: str) -> list[SafetySource]:
    matches = list(re.finditer(r"\[PubMed\s+(\d+)\]", context, re.I))
    sources: list[SafetySource] = []
    for index, match in enumerate(matches):
        end = matches[index + 1].start() if index + 1 < len(matches) else len(context)
        sources.append(SafetySource(label=f"PubMed {match.group(1)}", text=context[match.start() : end]))
    return sources
